src/scheduler.cc

```diff
@@ -267,7 +267,47 @@ Status SchedulerService::TaskInfo(ServerContext* context, const TaskInfoRequest*
   return Status::OK;
 }
 
-void SchedulerService::deliver_object_if_necessary(ObjRef canonical_objref, ObjStoreId from, ObjStoreId to) {
+Status SchedulerService::KillWorkers(ServerContext* context, const KillWorkersRequest* request, KillWorkersReply* reply) {
+  // TODO: Update reference counts
+  std::lock_guard<std::mutex>
+    failed_tasks_lock(failed_tasks_lock_),
+    get_queue_lock(get_queue_lock_),
+    computation_graph_lock(computation_graph_lock_),
+    fntable_lock(fntable_lock_),
+    avail_workers_lock(avail_workers_lock_),
+    task_queue_lock(task_queue_lock_),
+    workers_lock(workers_lock_);
+  // TODO: Fault handling... what if a worker refuses to die? We just assume it dies here.
+  for (size_t i = 0; i != workers_.size(); ++i) {
+    ClientContext client_context;
+    DieRequest die_request;
+    DieReply die_reply;
+    workers_[i].worker_stub->Die(&client_context, die_request, &die_reply);
+    TaskStatus status;
+    status.set_operationid(workers_[i].current_task);
+    status.set_worker_address(workers_[i].worker_address);
+    status.set_error_message("Worker killed by KillWorkers call in scheduler while executing task.");
+    failed_tasks_.push_back(status);
+    workers_[i].worker_stub.reset();
+  }
+  avail_workers_.clear();
+  fntable_.clear();
+  get_queue_.clear();
```

Comment:
There might be tasks still running that are waiting for objects, so I don't think we can clear the `get_queue` here.